src/core/middlewares/util/util.go

```diff
@@ -111,22 +129,104 @@ type MfInfo struct {
 	Tag        string
 	Digest     string
 
-	// Exist is to index the existing of the manifest in DB. If false, it's an new image for uploading.
-	Exist bool
+	References []distribution.Descriptor
+	Descriptor distribution.Descriptor
+
+	// manifestExist is to index the existing of the manifest in DB by (repository, digest)
+	manifestExist     bool
+	manifestExistErr  error
+	manifestExistOnce sync.Once
+
+	// artifact the artifact indexed by (repository, tag) in DB
+	artifact     *models.Artifact
+	artifactErr  error
+	artifactOnce sync.Once
+
+	ExclusiveBlobs []*models.Blob
+}
+
+// MutexKey returns mutex key of the manifest
+func (info *MfInfo) MutexKey(suffix ...string) string {
```

Comment:
I remember that we need to lock tag for count usage on intercept the put manifest request?

Reply:
Fixed